### PR TITLE
Fix consumed_valid_samples counting for several valid dataloaders

### PR DESCRIPTION
**Pull request by TevenLeScao**
Currently, the valid dataloaders accumulate the number of valid steps elapsed together: a single `args.consumed_valid_samples` counter is incremented for every dataset. This is an issue as each valid data sampler takes `args.consumed_valid_samples` as the elapsed valid steps to compare with the size of the dataset: it quickly grows bigger, which [breaks the data sampler initialization](https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/main/megatron/data/data_samplers.py#L73).

As every validation dataset is evaluated on equally, this PR fixes that issue by passing the correct number of elapsed valid steps per valid dataset, which is an equal fraction of `args.consumed_valid_samples for each data sampler.

### CODE REVIEW

**Comment by TevenLeScao** [comment]
Just realized the commit message was wrong, sorry!

**Review by stas00 — APPROVED** [review]
This makes sense.

I'm curious wrt to the code you linked to - why do validation samples ever run out? It doesn't train on those so if they start repeating - what practical difference does it have?

**Comment by TevenLeScao** [comment]
As per @jaredcasper :

> That sampler is only designed to go through the dataset once. If consumed_samples > total_samples then the iterator breaks (see line 93). Not necessarily fundamental, just how it got written.

Actually a dataloader with exactly enough points in it for all of the validation passes gets built, then gets read through once. I hesitated to switch to a `cyclic` dataloader, but figured this would be truer to what's actually happening.

**Comment by stas00** [comment]
If I understand things correctly, then the cyclic dataloader is the way to go. 

Especially since with ML some languages might not have too much data to spare.

**Comment by jaredcasper** [comment]
> Actually a dataloader with exactly enough points in it for all of the validation passes gets built, then gets read through once. I hesitated to switch to a cyclic dataloader, but figured this would be truer to what's actually happening.

Right. Sorry about the confusion, by "dataset" I meant the PyTorch dataset that we create, which has enough copies of the underlying data set in it to cover the number of iterations requested at the beginning of training. So the sampler only needs to go through it once.

**Comment by jaredcasper** [comment]
> If I understand things correctly, then the cyclic dataloader is the way to go. Especially since with ML some languages might not have too much data to spare.

We've found the best way to go through the base data set multiple times is to do what the Megatron dataset does, which is lay out the data N times, then do a global shuffle. When you go through the whole dataset once, then cycle back through it, even in a different oder, the training does not go as smoothly.

**Comment by stas00** [comment]
> When you go through the whole dataset once, then cycle back through it, even in a different order, the training does not go as smoothly.

But we are talking only validation here. Does the same answer still apply?

**Comment by jaredcasper** [comment]
Ah yes, sorry. Not sure how it would impact validation. We've always just used the same code for both.

**Comment by stas00** [comment]
Thank you for clarifying.

It's a specific issue with some languages having too little data, so we want as much as possible to go for training.

@TevenLeScao, I guess the other solution is to simply copy the validation data several times and then not needing to change the current machinery.

**Comment by TevenLeScao** [comment]
Ah, but that is already what is happening ! The validation dataset is copied at the onset of training until it's big enough for all the evals that are planned (so if you have 80 eval passes of 100 steps, it'll be size 8000 whatever the size of the starting dataset).
